### PR TITLE
Get user devices generic

### PR DESCRIPTION
**Pull request by tomer-pan**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-6611

## Description
Generic playbooks for retrieving user devices by email/username.

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [x] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

### CODE REVIEW

File: Packs/CommonPlaybooks/Playbooks/playbook-Get_User_Devices_by_Email_Address_-_Generic_README.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+| JAMF.MobileDevice.email_address | The mobile device user email address. | unknown |
+| JAMF.MobileDevice.room | The mobile device room. | unknown |
+| JAMF.MobileDevice.position | The mobile device position. | unknown |
+| JAMF.MobileDevice.building | The mobile device building. | unknown |
+| JAMF.MobileDevice.building_name | The mobile device building name. | unknown |
+| JAMF.MobileDevice.department | The mobile device department. | unknown |
+| JAMF.MobileDevice.department_name | The mobile device department name. | unknown |
+| ServiceNow.Computer | ServiceNow computer object. | unknown |
+| ServiceNow.Computer.ID | Computer system ID. | unknown |
+| ServiceNow.Computer.AssetTag | Computer Asset tag. | unknown |
```
Comment (ShirleyDenkberg) [inline Packs/CommonPlaybooks/Playbooks/playbook-Get_User_Devices_by_Email_Address_-_Generic_README.md:153]:
```suggestion
| ServiceNow.Computer.AssetTag | Computer asset tag. | unknown |
```

File: Packs/CommonPlaybooks/Playbooks/playbook-Get_User_Devices_by_Username_-_Generic_README.md [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+| JAMF.Computer.building_name | The computer building name. | unknown |
+| JAMF.Computer.department | The computer department. | unknown |
+| JAMF.Computer.department_name | The computer department name. | unknown |
+| JAMF.MobileDevice | Mobile device object. | unknown |
+| JAMF.MobileDevice.id | The mobile device ID. | unknown |
+| JAMF.MobileDevice.name | The mobile device name. | unknown |
+| JAMF.MobileDevice.udid | The mobile device UDID. | unknown |
+| JAMF.MobileDevice.serial_number | The mobile device serial number. | unknown |
+| JAMF.MobileDevice.mac_address | The mobile device MAC address. | unknown |
+| JAMF.MobileDevice.wifi_mac_address | The mobile device WIFI MAC address. | unknown |
```
Comment (ShirleyDenkberg) [inline Packs/CommonPlaybooks/Playbooks/playbook-Get_User_Devices_by_Username_-_Generic_README.md:144]:
```suggestion
| JAMF.MobileDevice.wifi_mac_address | The mobile device WI-FI MAC address. | unknown |
```

File: Packs/CommonPlaybooks/Playbooks/playbook-Get_User_Devices_by_Email_Address_-_Generic_README.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+| GSuite.MobileDevices.MobileListObjects.name | A list of the owner's usernames. | unknown |
+| GSuite.MobileDevices.MobileListObjects.email | A list of the owner's email addresses. | unknown |
+| GSuite.MobileDevices.MobileListObjects.model | The mobile device's model name. | unknown |
+| GSuite.MobileDevices.MobileListObjects.os | The mobile device's operating system. | unknown |
+| GSuite.MobileDevices.MobileListObjects.type | The type of mobile device. | unknown |
+| GSuite.MobileDevices.MobileListObjects.status | The mobile device's status. | unknown |
+| GSuite.MobileDevices.MobileListObjects.hardwareId | The IMEI/MEID unique identifier for Android hardware. | unknown |
+| GSuite.MobileDevices.MobileListObjects.firstSync | The date and time the mobile device was initially synchronized with the policy settings in the Admin console. | unknown |
+| GSuite.MobileDevices.MobileListObjects.lastSync | The date and time the mobile device was last synchronized with the policy settings in the Admin console. | unknown |
+| GSuite.MobileDevices.MobileListObjects.userAgent | Gives information about the mobile device such as the operating system version. | unknown |
```
Comment (ShirleyDenkberg) [inline Packs/CommonPlaybooks/Playbooks/playbook-Get_User_Devices_by_Email_Address_-_Generic_README.md:82]:
```suggestion
| GSuite.MobileDevices.MobileListObjects.userAgent | Information about the mobile device such as the operating system version. | unknown |
```

File: Packs/CommonPlaybooks/Playbooks/playbook-Get_User_Devices_by_Username_-_Generic_README.md [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+| GSuite.MobileDevices.MobileListObjects.name | A list of the owner's usernames. | unknown |
+| GSuite.MobileDevices.MobileListObjects.email | A list of the owner's email addresses. | unknown |
+| GSuite.MobileDevices.MobileListObjects.model | The mobile device's model name. | unknown |
+| GSuite.MobileDevices.MobileListObjects.os | The mobile device's operating system. | unknown |
+| GSuite.MobileDevices.MobileListObjects.type | The type of mobile device. | unknown |
+| GSuite.MobileDevices.MobileListObjects.status | The mobile device's status. | unknown |
+| GSuite.MobileDevices.MobileListObjects.hardwareId | The IMEI/MEID unique identifier for Android hardware. | unknown |
+| GSuite.MobileDevices.MobileListObjects.firstSync | The date and time the mobile device was initially synchronized with the policy settings in the Admin console. | unknown |
+| GSuite.MobileDevices.MobileListObjects.lastSync | The date and time the mobile device was last synchronized with the policy settings in the Admin console. | unknown |
+| GSuite.MobileDevices.MobileListObjects.userAgent | Gives information about the mobile device such as the operating system version. | unknown |
```
Comment (ShirleyDenkberg) [inline Packs/CommonPlaybooks/Playbooks/playbook-Get_User_Devices_by_Username_-_Generic_README.md:86]:
```suggestion
| GSuite.MobileDevices.MobileListObjects.userAgent | Information about the mobile device such as the operating system version. | unknown |
```

File: Packs/CommonPlaybooks/Playbooks/playbook-Get_User_Devices_by_Username_-_Generic_README.md [hunk 5]
```diff
@@ -0,0 +1,225 @@
+This playbook retrieves information on all of the associated user devices, based on the user's username.
+In order to get a generic output, the following information on all of the retrieved devices will be saved under the `UserDevices` context key:
+- Name
+- Serial Number
+- ID
+- Type
+- OS
+- Integration
+
+Please note that not all of the supported integrations will be able to retrieve this information.
```
Comment (ShirleyDenkberg) [inline Packs/CommonPlaybooks/Playbooks/playbook-Get_User_Devices_by_Username_-_Generic_README.md:10]:
```suggestion
Note that not all of the supported integrations will be able to retrieve this information.
```

File: Packs/CommonPlaybooks/Playbooks/playbook-Get_User_Devices_by_Username_-_Generic_README.md [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+* servicenow-query-computers
+
+## Playbook Inputs
+
+---
+
+| **Name** | **Description** | **Default Value** | **Required** |
+| --- | --- | --- | --- |
+| Username | The username of the user. |  | Optional |
+| GsuiteCustomerID | In case of using Google Workspace, a customer ID is needed. |  | Optional |
```
Comment (ShirleyDenkberg) [inline Packs/CommonPlaybooks/Playbooks/playbook-Get_User_Devices_by_Username_-_Generic_README.md:62]:
```suggestion
| GsuiteCustomerID | If using Google Workspace, a customer ID is needed. |  | Optional |
```

File: Packs/CommonPlaybooks/ReleaseNotes/2_3_67.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+- This playbook retrieves information on all of the associated user devices, based on the user email.
+In order to get a generic output, the following information on all of the retrieved devices will be saved under the `UserDevices` context key:
+- Name
+- Serial Number
+- ID
+- Type
+- OS
+- Integration
+
+Please note that not all of the supported integrations will be able to retrieve this information.
```
Comment (ShirleyDenkberg) [inline Packs/CommonPlaybooks/ReleaseNotes/2_3_67.md:32]:
```suggestion
Note that not all of the supported integrations will be able to retrieve this information.
```

File: Packs/CommonPlaybooks/Playbooks/playbook-Get_User_Devices_-_Generic_README.md [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+- Name
+- Serial Number
+- ID
+- Type
+- OS
+- Integration
+
+Please note that not all of the supported integrations will be able to retrieve this information.
+
+In order to get the full list of supported integrations, please read the following sub-playbooks descriptions:
```
Comment (ShirleyDenkberg) [inline Packs/CommonPlaybooks/Playbooks/playbook-Get_User_Devices_-_Generic_README.md:12]:
```suggestion
In order to get the full list of supported integrations, read the following sub-playbooks descriptions:
```

**Comment by ShirleyDenkberg** [comment]
@idovandijk @AdiPeret @michalgold Doc review completed.

**Review by idovandijk — CHANGES_REQUESTED** [review]
Great playbooks. Looking forward to have these in content! You've put a lot of effort into this. I did find some small mistakes. My comments are numbered to make it easier for you to refer to them in your replies:

### Get User Devices by Username - Generic
I like the trick you did here :)
image

1. The SetMultipleValues task doesn't work correctly in my test:
image
please check why the "integration" field is empty. For testing you can use the data I sent you on Slack. Maybe you should use `SetAndHandleEmpty` like you did in the other integrations?

2. The Microsoft Defender Advanced Threat Protection integration is deprecated. Please check if there is an alternative that you can replace it with, or remove it.

3. Is it possible to get devices from Google without `GsuiteCustomerID`?

4. AD seems to be missing an "OS" output here, but I do see it in the `Get User Devices by Email Address` playbook

### Get User Devices by Email Address - Generic

5. Please add an Else path for MS Graph:
image


### Get User Devices - Generic
6. No need to append UserDevices. Task number 3 will only happen once both playbooks finished. You can simply `Set` the UserDevices and apply a `Unique` filter (without appending). Correct me if I'm wrong.

7. Please fix description - UserDevices is an object containing various data fields (not just name):
image

8. Maybe this is not fixable, but if we find the same data by username and by email, we would have somewhat duplicated data. This is just an FYI:
image


### General Comments
9. You are outputting "Type" for AD, but "Model" for all other integrations. Please fix all occurrences so your outputs are consistent with the outputs declared in the playbooks and with the description of the playbook:
image
image

Note that in the descriptions, outputs and AD you are using "Type". All the other integrations use "Model" in their SetAndHandleEmpty tasks. The change requires updating the declared outputs, playbook descriptions and READMEs.

10. Jamf is also outputting MAC address:
image

let's add it to the UserDevices object. Note that it requires the change in both playbooks, their declared UserDevices outputs (need to add UserDevices.MACAddress), and the descriptions of the 3 playbooks, as well as the readmes.

**Comment by tomer-pan** [comment]
@idovandijk All of your suggestions/comments were fixed and implemented.

**Review by idovandijk — APPROVED** [review]
GJ!

**Pull request merged**